### PR TITLE
Implemented .ru Viewer

### PR DESCRIPTION
**Pull request by feliwir**
- many additions to the Apt system
- Added triangle rendering to Graphics2D

### CODE REVIEW

**Comment by feliwir** [comment]
Here is an example for a displayed .ru file. 
image

Those shapes are later combined and animated in the .apt format.

File: src/OpenSage.DataViewer/OpenSage.DataViewer.csproj [hunk 1]
```diff
@@ -3,6 +3,7 @@
     <TargetFramework>net462</TargetFramework>
     <ApplicationIcon>Resources\AppIcon.ico</ApplicationIcon>
     <LangVersion>latest</LangVersion>
+    <Platforms>AnyCPU</Platforms>
```
Comment (tgjones) [inline src/OpenSage.DataViewer/OpenSage.DataViewer.csproj]: Is this change necessary?

File: src/OpenSage.Game/Content/AptLoader.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OpenSage.Content
+{
+    class AptLoader
```
Comment (tgjones) [inline src/OpenSage.Game/Content/AptLoader.cs]: Will this be filled in later? Should it be removed for now?

File: src/OpenSage.Game/Content/ShapeLoader.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+{
+    internal sealed class ShapeLoader : ContentLoader<ShapeComponent>
+    {
+        public ShapeLoader(ContentManager contentManager)
+        {
+            switch (contentManager.SageGame)
+            {
+                case SageGame.CncGenerals:
+                case SageGame.CncGeneralsZeroHour:
+                    contentManager.IniDataContext.LoadIniFile(@"Data\English\HeaderTemplate.ini");
```
Comment (tgjones) [inline src/OpenSage.Game/Content/ShapeLoader.cs]: Are header templates and mapped images really necesary for `.ru`? I guess this is left in from `WindowLoader`.

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 4]
```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+                    {
+                        //no data parsed yet
+                        case GeometryStyle.Undefined:
+                            break;
+                        case GeometryStyle.TexturedTri:
+                            geometry.Entries.Add(new GeometryTexturedTriangles(tris, color,image,rotMat,translation));
+                            tris.Clear();
+                            break;
+                        case GeometryStyle.SolidTri:
+                            geometry.Entries.Add(new GeometrySolidTriangles(tris, color));
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: I think it would be cleaner to pass `new List<Triangle2D>(tris)` here, rather than doing it in the `GeometrySolidTriangles` constructor.

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 5]
```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+                            tris.Clear();
+                            break;
+                        case GeometryStyle.Line:
+                            geometry.Entries.Add(new GeometryLines(lines, color, thickness));
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: Same here - I think it would be cleaner to pass `new List<Line2D>(lines)` here.

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 6]
```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+
+        private void CalculateBoundings()
+        {
+            //find topleft and botright point
+            var topLeft = new Vector2 { X = float.MaxValue, Y = float.MaxValue };
+            var botRight = new Vector2 { X = float.MinValue, Y = float.MinValue };
+
+            foreach (var entry in Entries)
+            {
+                if (entry is GeometryLines gl)
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: Nice use of C#'s new pattern syntax :) I think you could go one step further and use the new switch/case pattern matching:

``` csharp
switch (entry)
{
  case GeometryLines gl:
    break;

  case GeometrySolidTriangles gst:
    break;

  // etc.
}
```

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 7]
```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+                    }
+                }
+            }
+
+            var size = botRight - topLeft;
+            BoundingBox = new RectangleF(topLeft.X, topLeft.Y, size.X, size.Y);
+        }
+    }
+
+    public struct GeometrySolidTriangles : GeometryEntry
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: Any particular reason why this is a `struct`? Usually if a type involves inheritance, and reference fields (like the `List<Triangle2D>` below) then I would make it a `class`.

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 8]
```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+    {
+        public ColorRgba Color { get; private set; }
+        public List<Triangle2D> Triangles { get; private set; }
+
+        public GeometryStyle Type
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: You could use an expression-based property here:

``` csharp
public GeometryStyle Type => GeometryStyle.SolidTri;
```

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 9]
```diff
@@ -17,13 +43,273 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+            }
+
+            var size = botRight - topLeft;
+            BoundingBox = new RectangleF(topLeft.X, topLeft.Y, size.X, size.Y);
+        }
+    }
+
+    public struct GeometrySolidTriangles : GeometryEntry
+    {
+        public ColorRgba Color { get; private set; }
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: When the property is (only) set by the constructor, you can remove the setter:

``` csharp
public ColorRgba Color { get; }
```

File: src/OpenSage.Game/Data/Apt/ImageMap.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+namespace OpenSage.Data.Apt
+{
+
+    public interface ImageAssignment
+    {
+        bool HasBounds { get; }
+        int TextureId { get; }
+    }
+
+    public struct DirectAssignment : ImageAssignment
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/ImageMap.cs]: This should probably be a `class`, along with the other assignment types. Using a struct that implements an interface involves boxing, which is relatively slow.

File: src/OpenSage.Game/Data/Apt/ImageMap.cs [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenSage.Mathematics;
+
+namespace OpenSage.Data.Apt
+{
+
+    public interface ImageAssignment
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/ImageMap.cs]: Interfaces should be prefixed with "I", so please could this be `IImageAssignment`.

File: src/OpenSage.Game/Data/Apt/ImageMap.cs [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+            _textureId = texture;
+        }
+    }
+
+    public struct RectangleAssignment : ImageAssignment
+    {
+        private int _textureId;
+        private Rectangle _textureRect;
+
+        public bool HasBounds
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/ImageMap.cs]: These can all use expression-based property syntax (`public bool HasBounds => true`).

File: src/OpenSage.Game/Gui/Apt/AptSystem.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                //CreateSizeDependentResources(guiComponent);
+            }
+        }
+
+
+
+        private static void DoActionRecursive(UIElement element, Func<UIElement, bool> action)
+        {
+            if (!action(element))
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/AptSystem.cs]: This looks specific to `.wnd` and can be removed?

File: src/OpenSage.Game/Gui/Apt/ShapeComponent.cs [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            var posY = (int) Math.Round((viewportSize.Height - (shapeBounding.Height * ratio)) / 2);
+
+            return new Rectangle(posX, posY, newWidth, newHeight);
+        }
+
+        public void Layout(GraphicsDevice gd, in Size windowSize)
+        {
+            float _scale = 0.0f;
+
+            var Frame = CalculateFrame(Shape.BoundingBox, windowSize, out _scale);
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ShapeComponent.cs]: This variable should be named `frame`.

File: src/OpenSage.Game/Gui/Apt/ShapeComponent.cs [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+                new SpriteVertex(new Vector2(left, bottom * -1), new Vector2(0, 1))
+            };
+            _vertexBuffer = Buffer<SpriteVertex>.CreateStatic(
+                gd,
+                vertices,
+                BufferBindFlags.VertexBuffer);
+
+        }
+
+        private ColorRgbaF ToFloatColor(ColorRgba color)
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ShapeComponent.cs]: This already exists, as the incorrectly-named `ColorRgba.ToColorRgba()` method. Feel free to use that method and fix its name :)

File: src/OpenSage.Game/Gui/Apt/ShapeComponent.cs [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        {
+            using (var drawingContext = new DrawingContext(HostPlatform.GraphicsDevice2D, _texture))
+            {
+                drawingContext.Begin();
+
+                drawingContext.Clear(new ColorRgbaF(0,0,0,1));
+
+                foreach (var e in Shape.Entries)
+                {
+                    if(e is GeometryLines l)
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ShapeComponent.cs]: This could be replaced with a switch/case:

``` csharp
switch (entry)
{
  case GeometryLines l:
    break;

  // etc.
}
```

File: src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs [hunk 17]
```diff
@@ -16,7 +19,8 @@ private void PlatformConstruct(GraphicsDevice2D graphicsDevice, Texture targetTe
 
             var dxgiSurface = targetTexture.DeviceResource.QueryInterface<Surface>();
 
-            _bitmap = AddDisposable(new Bitmap1(graphicsDevice.DeviceContext, dxgiSurface));
+           _bitmap = AddDisposable(new Bitmap1(graphicsDevice.DeviceContext, dxgiSurface));
```
Comment (tgjones) [inline src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs]: Somehow we lost a space here...

File: src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs [hunk 18]
```diff
@@ -84,16 +100,83 @@ private void PlatformFillRectangle(in RawRectangleF rect, in ColorRgbaF fillColo
...
+                sink.SetFillMode(SharpDX.Direct2D1.FillMode.Winding);
+                sink.AddLine(ToRawVector2(tri.X2, tri.Y2));
+                sink.AddLine(ToRawVector2(tri.X3, tri.Y3));
+                sink.EndFigure(FigureEnd.Closed);
+                sink.Close();
+                _graphicsDevice.DeviceContext.FillGeometry(geometry, brush);
+            }
+        }
+
+        private void PlatformFillTriangle(in RawTriangleF tri, in Texture tex, in RawMatrix3x2 transform)
```
Comment (tgjones) [inline src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs]: Please remove the `in` modifier from the `tex` parameter - it's only useful on `struct` parameters. For reference parameters it's already the default. And please could you use full names, like `texture` instead of `tex`?

File: src/OpenSage.LowLevel/Graphics2D/DrawingContext.cs [hunk 19]
```diff
@@ -34,6 +34,21 @@ public void DrawRectangle(in RawRectangleF rect, in ColorRgbaF strokeColor, floa
...
+        {
+            PlatformDrawLine(line, strokeColor);
+        }
+
+        public void FillTriangle(in RawTriangleF tri, in ColorRgbaF fillColor)
+        {
+            PlatformFillTriangle(tri, fillColor);
+        }
+
+        public void FillTriangle(in RawTriangleF tri, in Texture texture,RawMatrix3x2 transform)
```
Comment (tgjones) [inline src/OpenSage.LowLevel/Graphics2D/DrawingContext.cs]: Please remove the `in` modifier from the `texture` parameter.

File: src/OpenSage.sln [hunk 20]
```diff
@@ -13,34 +13,56 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenSage.DataViewer", "Open
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenSage.DataViewer.Windows", "OpenSage.DataViewer.Windows\OpenSage.DataViewer.Windows.csproj", "{1D432BD6-B14F-4008-88B3-E3DC52AC7380}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "OpenSage.LowLevel", "OpenSage.LowLevel\OpenSage.LowLevel.csproj", "{AA6CCAEB-CC90-4463-A67E-336580BCD036}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "OpenSage.LowLevel", "OpenSage.LowLevel\OpenSage.LowLevel.csproj", "{AA6CCAEB-CC90-4463-A67E-336580BCD036}"
```
Comment (tgjones) [inline src/OpenSage.sln]: Are any of these `.sln` changes necessary? If not please could you revert them?

File: src/OpenSage.DataViewer/OpenSage.DataViewer.csproj (hunk 1 above)
Comment (feliwir) [reply src/OpenSage.DataViewer/OpenSage.DataViewer.csproj]: No idea where it came from

File: src/OpenSage.Game/Content/AptLoader.cs (hunk 2 above)
Comment (feliwir) [reply src/OpenSage.Game/Content/AptLoader.cs]: Will be filled

File: src/OpenSage.Game/Content/ShapeLoader.cs (hunk 3 above)
Comment (feliwir) [reply src/OpenSage.Game/Content/ShapeLoader.cs]: Yes

File: src/OpenSage.Game/Data/Apt/Geometry.cs (hunk 4 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Apt/Geometry.cs]: Will do

File: src/OpenSage.Game/Data/Apt/Geometry.cs (hunk 6 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Apt/Geometry.cs]: Wow, that's really amazing 👍

File: src/OpenSage.Game/Data/Apt/Geometry.cs (hunk 7 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Apt/Geometry.cs]: I thought struct's might be faster/smaller/better. Will change

File: src/OpenSage.Game/Data/Apt/ImageMap.cs (hunk 11 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Apt/ImageMap.cs]: Yup

File: src/OpenSage.sln (hunk 20 above)
Comment (feliwir) [reply src/OpenSage.sln]: Not sure where those came from, will change

File: src/OpenSage.Game/Data/Apt/Geometry.cs [hunk 21]
```diff
@@ -17,13 +43,251 @@ public static Geometry FromFileSystemEntry(FileSystemEntry entry)
...
+                        case GeometryStyle.TexturedTri:
+                            geometry.Entries.Add(new GeometryTexturedTriangles(new List<Triangle2D>(tris), color, image, rotMat, translation));
+                            tris.Clear();
+                            break;
+                        case GeometryStyle.SolidTri:
+                            geometry.Entries.Add(new GeometrySolidTriangles(new List<Triangle2D>(tris), color));
+                            tris.Clear();
+                            break;
+                        case GeometryStyle.Line:
+                            geometry.Entries.Add(new GeometryLines(lines, color, thickness));
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Geometry.cs]: This should be `new List<Line2D>(lines)`, instead of doing it inside the constructor?

File: src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs [hunk 22]
```diff
@@ -84,16 +99,76 @@ private void PlatformFillRectangle(in RawRectangleF rect, in ColorRgbaF fillColo
...
         {
             return new SharpDX.Mathematics.Interop.RawRectangleF(value.X, value.Y, value.X + value.Width, value.Y + value.Height);
         }
 
+        private static SharpDX.Mathematics.Interop.RawMatrix3x2 ToRawMatrix3x2(in RawMatrix3x2 m)
+        {
+            return new SharpDX.Mathematics.Interop.RawMatrix3x2(m.M11, m.M12, m.M21, m.M22, m.M31, m.M32);
+        }
+
+        private static SharpDX.Mathematics.Interop.RawVector2 ToRawVector2(in float x, in float y)
```
Comment (tgjones) [inline src/OpenSage.LowLevel/Graphics2D/DrawingContext.D2D.cs]: No need for `in` on primitive types - on a 64-bit platform, it will actually end up using *more* memory (64-bit pointer vs 32-bit `float`).

**Comment by tgjones** [comment]
Thank you for this! Looking forward to the next one :)
